src/control/control.rs

```diff
@@ -0,0 +1,173 @@
+                    match self.k.convert(keyp) {
+                        Action::Up => {
+                            self.u = if self.u == Keystate::Hold {Keystate::None} else {Keystate::Tap};
+                        },
+                        Action::Down => {
+                            self.d = if self.d == Keystate::Hold {Keystate::None} else {Keystate::Tap};
+                        },
+                        Action::Left => {
+                            self.l = if self.l == Keystate::Hold {Keystate::None} else {Keystate::Tap};
+                        },
+                        Action::Right => {
+                            self.r = if self.r == Keystate::Hold {Keystate::None} else {Keystate::Tap};
+                        },
+                        _ => {}
+                    }
+                },
+                _ => {}
+            };
+        }
+
+        match self.u {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.u = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        match self.d {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.d = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        match self.l {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.l = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        match self.r {
+            Keystate::Press(tim) => {
+                if now > tim + KEYHOLDDELAY {
+                    self.r = Keystate::Hold;
+                }
+            }
+            _ => {}
+        }
+        
+        let dir = match (self.u, self.d, self.l, self.r) {
+            (Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None, Keystate::None) => Some((0, -1)),
+            (Keystate::None, Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None) => Some((0, 1)),
+            (Keystate::None, Keystate::None, Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, 0)),
+            (Keystate::None, Keystate::None, Keystate::None, Keystate::Hold | Keystate::Tap) => Some((1, 0)),
+            
+            (Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, -1)),
+            (Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap) => Some((1, -1)),
+            (Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None) => Some((-1, 1)),
+            (Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap, Keystate::None, Keystate::Press(_) | Keystate::Hold | Keystate::Tap) => Some((1, 1)),
```

Comment:
https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-remaining-parts-of-a-value-with-
This would help you shorten a couple lines here. What I would actually do is have the 8 directions first, and have the 4 directions after it and make liberal use of `..` just to reduce clutter.

Reply:
I don't want to ignore remaining parts of a value, that creates poorly defined behaviour when 3 or 4 buttons are pressed

Reply:
This is reasonable, for later changes